components/script/dom/htmliframeelement.rs

```diff
@@ -287,7 +287,24 @@ impl HTMLIFrameElement {
 
         let url = self.get_url();
 
-        // TODO: check ancestor browsing contexts for same URL
+        // return early if there's an ancestor browsing context
+        // "whose active document's url, ignoring fragments, is equal"
+        let mut ancestor = window.GetParent();
+        while let Some(a) = ancestor {
```

Comment:
This implementation is currently limited to same-origin iframes; unfortunately we need to query the constellation for the full iframe ancestor chain to be certain.

Reply:
Thanks for pointing that out. Could you give me a rough sketch of how that goes?

Reply:
1. add a new message to https://github.com/servo/servo/blob/688d0768898343347b7b5ba140cbc6340677be90/components/script_traits/script_msg.rs#L117 that contains an IpcSender that will be used to send the response, along with the desired url and the frame's BrowsingContextId
1. add a message handler to https://github.com/servo/servo/blob/688d0768898343347b7b5ba140cbc6340677be90/components/constellation/constellation.rs#L1688 that checks the `browsing_context` map, then gets the browsing context's parent pipeline id (https://github.com/servo/servo/blob/688d0768898343347b7b5ba140cbc6340677be90/components/constellation/browsingcontext.rs#L63) and uses that to get the pipeline from `pipelines`, which contains a browsing context id which corresponds with the parent. Repeat until there are no parents.
1. send the result of checking the ancestors on the provided ipc sender.
1. in the script thread code, use the global scope's `script_to_constellation_chan` to send the new message, first creating a new ipc channel to include in it
1. wait on receiving a message on the new ipc channel receiver

Reply:
Does checking the url of each browsing context have to happen on that browsing context's own script thread?

Reply:
Are you asking if the check could happen as part of navigation in the constellation instead? I don't know the answer without reading the spec, sadly.

Reply:
What I'm not sure about is how to ask a browsing context "what's your URL" without crossing a memory boundary I'm not supposed to cross. This problem might be better tackled by someone who's gone deeper into Servo's threading systems.

Reply:
Simplifying realization: if it's not the same script thread that asked the question, it must also not be the same URL, since within an origin we're sticking to one thread. Constellation could hand back a list of contexts, then script thread could iterate them to ask the question of the same-origin ones.

Reply:
I can imagine the list of parents changing in mid-check due to something happening in another script thread, like if we're in a complicated tree of iframes and a cross-origin parent hands us off for adoption by another iframe that's also in that origin. The implied synchronousness of this ancestry-check operation in the spec doesn't really match the way Servo does things.

Reply:
Ick, yeah. I think I see what you're getting at. One solution might be for the constellation to reply with a list of channels to ancestor script threads, and the original script thread can then query those threads directly.

Reply:
Does the cross-origin situation apply here? As I understand it, here we are operating on the iframe element, the BC container, not the actual nested BC which potentially is cross-site in which case it would be running in another process. 
 
(and if the nested BC that will be loaded as a result of this navigation is cross-site, I think this implies by definition that the parent BC does not have the same url?)

Reply:
Ah you mean if the nested BC is cross-site and then loads another iframe for that is cross-site to it, but matches the url of the BC where the first iframe is contained within? Yeah that's a tough one.

Reply:
I think you might want to not completely follow the spec to the letter here. So instead of returning, you could not go ahead with the navigation when the `ScriptMsg::ScriptLoadedURLInIFrame` is handled by the constellation, using the `old_pipeline_id` field to check the url of all `Pipeline` that are ancestor to it? 

So you'd have to essentially to the same check as here, but in the constellation over at https://github.com/servo/servo/blob/89384ffec35529571be41882d77d4b0438c1f8b9/components/constellation/constellation.rs#L2940

I think it could be done in a separate PR, since this code can be kept for when all parents happen to be running in the same script-thread? This would be like an early return case, sort of independent of a later check happening in the constellation?

So I think the "local" check you perform here is the "early abort" case that matches the spec. And then in the cross-site case, you'd have to do the check a bit later when the navigation is handled by the constellation. Since the spec only returns, I think the effect of the delay will not be noticeable to script (unloading only happens later if the navigation goes through, I think).

Reply:
One last thing, if the navigation were to be disallowed in the constellation, you'd probably have to send a message to the script-thread to undo some of the changes that occur as part of https://github.com/servo/servo/blob/89384ffec35529571be41882d77d4b0438c1f8b9/components/script/dom/htmliframeelement.rs#L109 when the navigation is initiated.